File: mac/Tests/AikoKitTests/ClaudeInstallTests.swift
import Testing

@testable import AikoKit

struct ClaudeInstallTests {
    static let home = #"C:\Users\someone"#

    @Test
    func freshPathJoinsMachineAndUserAndExpandsThem() {
        let path = ClaudeInstall.freshPath(
            .windows,
            machinePath: #"C:\Windows"#,
            userPath: #"%USERPROFILE%\.local\bin"#,
            expand: { $0.replacingOccurrences(of: "%USERPROFILE%", with: Self.home) })

        #expect(path == #"C:\Windows;C:\Users\someone\.local\bin"#)
        #expect(
            ClaudeInstall.freshPath(.windows, machinePath: #"C:\Windows"#, userPath: nil, expand: { $0 })
                == #"C:\Windows"#)
    }

    @Test
    func findsClaudePastTheCommandFolder() {
        let files: Set<String> = [#"C:\aiko\bin\claude.exe"#, #"C:\tools\claude.exe"#]

        #expect(
            ClaudeInstall.find(
                .windows, freshPath: #"C:\aiko\bin;C:\tools"#, commandFolder: #"C:\aiko\bin"#,
                userProfile: Self.home, exists: files.contains) == #"C:\tools\claude.exe"#)
    }

    @Test
    func fallsBackToTheNativeInstallerFolder() {
        let native = Self.home + #"\.local\bin\claude.exe"#

        #expect(
            ClaudeInstall.find(
                .windows, freshPath: #"C:\Windows"#, commandFolder: #"C:\aiko\bin"#,
                userProfile: Self.home, exists: { $0 == native }) == native)
        #expect(
            ClaudeInstall.find(
                .windows, freshPath: #"C:\Windows"#, commandFolder: #"C:\aiko\bin"#,
                userProfile: Self.home, exists: { _ in false }) == nil)
    }

    @Test(arguments: [
        ("Work", #"C:\Users\someone\.claude-work"#),
        ("Основная", #"C:\Users\someone\.claude-osnovnaya"#),
        ("!!", #"C:\Users\someone\.claude-env"#),
    ] as [(String, String)])
    func aNewFolderIsNamedAfterTheEnvironment(name: String, folder: String) {
        #expect(
            ClaudeInstall.newConfigFolder(
                .windows, environmentName: name, userProfile: Self.home, folderExists: { _ in false })
                == folder)
    }

    @Test
    func aTakenFolderNameGetsANumber() {
        let taken: Set<String> = [Self.home + #"\.claude-work"#, Self.home + #"\.claude-work-2"#]

        #expect(
            ClaudeInstall.newConfigFolder(
                .windows, environmentName: "Work", userProfile: Self.home,
                folderExists: { path in taken.contains(where: { $0.caseInsensitiveCompare(path) == .orderedSame }) })
                == Self.home + #"\.claude-work-3"#)
    }
}

File: tests/Aiko.Core.Tests/ClaudeInstallTests.cs
using Aiko.Core;

namespace Aiko.Core.Tests;

public class ClaudeInstallTests
{
    private const string Home = @"C:\Users\someone";

    [Fact]
    public void Fresh_path_joins_machine_and_user_and_expands_them()
    {
        var path = ClaudeInstall.FreshPath(Windows,
            @"C:\Windows",
            @"%USERPROFILE%\.local\bin",
            text => text.Replace("%USERPROFILE%", Home));

        Assert.Equal(@"C:\Windows;C:\Users\someone\.local\bin", path);
        Assert.Equal(@"C:\Windows", ClaudeInstall.FreshPath(Windows, @"C:\Windows", null, t => t));
    }

    [Fact]
    public void Finds_claude_past_the_command_folder()
    {
        var files = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            @"C:\aiko\bin\claude.exe", @"C:\tools\claude.exe",
        };

        Assert.Equal(@"C:\tools\claude.exe", ClaudeInstall.Find(Windows, @"C:\aiko\bin;C:\tools", @"C:\aiko\bin", Home, files.Contains));
    }

    [Fact]
    public void Falls_back_to_the_native_installer_folder()
    {
        var native = Home + @"\.local\bin\claude.exe";

        Assert.Equal(native, ClaudeInstall.Find(Windows, @"C:\Windows", @"C:\aiko\bin", Home, p => p == native));
        Assert.Null(ClaudeInstall.Find(Windows, @"C:\Windows", @"C:\aiko\bin", Home, _ => false));
    }

    [Theory]
    [InlineData("Work", @"C:\Users\someone\.claude-work")]
    [InlineData("Основная", @"C:\Users\someone\.claude-osnovnaya")]
    [InlineData("!!", @"C:\Users\someone\.claude-env")]
    public void A_new_folder_is_named_after_the_environment(string name, string folder)
    {
        Assert.Equal(folder, ClaudeInstall.NewConfigFolder(Windows, name, Home, _ => false));
    }

    [Fact]
    public void A_taken_folder_name_gets_a_number()
    {
        var taken = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            Home + @"\.claude-work", Home + @"\.claude-work-2",
        };

        Assert.Equal(Home + @"\.claude-work-3", ClaudeInstall.NewConfigFolder(Windows, "Work", Home, taken.Contains));
    }
}
